File: trees/cppRealizations/binaryTreeRightSideView.cpp
#include "btf.h"
#include <vector>
#include <queue>


class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        deque<TreeNode*> q = {root};
        while(!q.empty()){
            TreeNode* rightSide = nullptr;
            for(int i = 0; i < q.size(); i++){
                TreeNode* node = q.front();
                q.pop_front();
                if(node != nullptr){
                    rightSide = node;
                }
                if(node->left != nullptr){
                    q.push_back(node->left);
                }
                if(node->right != nullptr){
                    q.push_back(node->right);
                }
            }
            if(rightSide != nullptr){
                res.push_back(rightSide->val);
            }
        }
        return res;
    }
};


int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(5);
    root->right->right = new TreeNode(4);
    cout << "Input:" << endl;
    btf(root);
    cout << "Solution:" << endl;
    vector<int> result = Solution().rightSideView(root);
    cout << "[ ";
    for(auto item : result){
        cout << item << ", ";
    }
    cout << "]" << endl;
    return 0;
}

File: trees/binaryTreeRightSideView.py
from trees.btf import *


class Solution:
    def __init__(self):
        self.hash = {}

    def rightSideViewMySolution(self, root: TreeNode) -> list[int]:
        def dfs(node, level: int, side: str):
            if not node:
                return
            if level not in self.hash:
                self.hash[level] = [(node.val, side)]
                # self.hash[level] = [side]
            else:
                self.hash[level].append((node.val,side))
                # self.hash[level].append(side)
            dfs(node.right, level + 1, 'r')
            dfs(node.left, level + 1, 'l')

        dfs(root, 0, 'r')
        result = []
        for i in self.hash.keys():
            result.append(self.hash[i][0][0])
        return result
    
    def rightSideView(self, root: TreeNode) -> list[int]:
        res = []
        q = deque([root])
        while q:
            rightSide = None
            for i in range(len(q)):
                node = q.popleft()
                if node:
                    rightSide = node
                    q.append(node.left)
                    q.append(node.right)
            if rightSide:
                res.append(rightSide.val)
        return res





if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(5)
    root.right.right = TreeNode(4)

    # root = TreeNode(1)
    # root.left = TreeNode(2)

    print("Input:")
    btf(root)
    print("Sulution:")
    print("MySolution")
    MySolution = Solution().rightSideViewMySolution(root)
    print(MySolution)
    print("NeetCode")
    NeetCode = Solution().rightSideView(root)
    print(NeetCode)
